Implement ignoring directories based on wildcards.

lberki opened the pull request:
This is accomplished by a new directive in REPO.bazel, "ignore_directories()". It takes a single argument, a list of directories to ignore and it allows the same wildcards as glob().

This is done separately from .bazelignore to provide a migration path off of that weird single-purpose configuration file.

Implementing this requires splitting RepoFileFunction into two: a part that parses the repository file and one that creates a PackageArgs instance. This was necessary to avoid a Skyframe dependency cycle: when a WORKSPACE file is present and it loads a .bzl file from a repository with a REPO.bazel file, the repo mapping for the main repository depends on the WORKSPACE file, which depends on the .bzl file, which depends on the IgnoredPackagePrefixesValue of its repository, which then depends on the repo mapping of the main repository and the one the .bzl file is in, which then depend on the WORKSPACE file.

Fixes #7093.

RELNOTES[NEW]: REPO.bazel now allows another directive, "ignore_directories()". It takes a list of directories to ignore just like .bazelignore does, but with glob semantics.

meteorcloudy commented:
Thanks! Waiting for @Wyverald to take a look!

fmeum commented on src/main/java/com/google/devtools/build/lib/cmdline/IgnoredSubdirectories.java:
```diff
@@ -29,42 +41,98 @@
  * <p>This is currently just a prefix, but will eventually support glob-style wildcards.
```
Can be updated now that globs are supported.

lberki replied:
Done (not exported yet, though, stay tuned)

fmeum commented on src/main/java/com/google/devtools/build/lib/vfs/UnixGlob.java:
```diff
@@ -901,29 +901,56 @@ public static void removeExcludes(Set<String> paths, Collection<String> excludes
     }
     if (j == path.length) {
-      return false;
+      return matchMode == MatchMode.CAN_MATCH_CHILD ? true : false;
```
```suggestion
      return matchMode == MatchMode.CAN_MATCH_CHILD;
```

lberki replied:
Indeed, rookie mistake :)

Wyverald commented on src/main/java/com/google/devtools/build/lib/packages/RepoCallable.java:
since this file no longer just contains the "repo callable", it should be renamed to `RepoFileGlobals` following other examples (`ModuleFileGlobals`, etc)

Wyverald commented on src/main/java/com/google/devtools/build/lib/packages/RepoCallable.java:
```diff
@@ -27,6 +29,25 @@ private RepoCallable() {}
+              @ParamType(type = Sequence.class, generic1 = String.class),
+            })
+      })
+  public Object ignoreDirectories(Iterable<?> dirsUnchecked, StarlarkThread thread)
```
does just returning void not work?

lberki replied:
Indeed, I just copy-pasted it from `repoCallable()`. It works alright.

Wyverald reviewed (commented):
so I reviewed some more superficial parts of this change before happening on to the whole "splitting RepoFileFunction into two parts" thing. After reading the context in https://github.com/bazelbuild/bazel/issues/7093#issuecomment-2418821279, my initial gut reaction is that this is a lot of extra complexity that's really not justified by its gains.

Besides Fabian's suggestion to cut edge 8 (which you had a counterargument for), I was also thinking if we could just cut edge 1 -- that is, WORKSPACE is going away anyway, so let's not worry about that. Essentially, we can try to make the assertion that computing the main repo mapping does _not_ transitively depend on `IgnoredPackagePrefixesValue`.

Unfortunately, that's not true... If we have a non-registry override for a `bazel_dep`, and the override refers to a patch file (which is addressed by a label), then we'll need to look up the package for that label. (Not to mention we actually need to load `@bazel_tools//tools/build_defs/repo:http.bzl` anyway, but I was hoping those could be avoided using exceptions -- which we already have quite a few of, regarding stuff in `@bazel_tools`.)

So all that was a long way to say I haven't finished the review yet, and will get to the meaty parts now.

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/RepoFileFunction.java:
```diff
@@ -87,29 +89,23 @@ public SkyValue compute(SkyKey skyKey, Environment env)
+    return evalRepoFile(
+        repoFile,
+        repoName,
+        RepositoryMapping.ALWAYS_FALLBACK,
```
actually, how is this split helping things? It looks like we end up always using `null` for the main repo mapping, which ends up being the same as cutting edge 8. So the split is actually not even doing anything?

(I expected the split to be done in a way that runs the REPO.bazel file twice: the first time without a main repo mapping just to get the ignore list, and the second time _with_ the main repo mapping to get the repo's PackageArgs. If we're not running the REPO.bazel file twice, there doesn't seem to be any point in splitting.)

Wyverald replied:
I see that, but my point is that netiher the first nor the second function is actually using the **main** repo mapping in this PR. This means we wouldn't have needed the split in the first place (IIUC the cycle was caused by the _main_ repo mapping, not the mapping of the repo where REPO.bazel is being processed). IOW, why can't IgnoredPrefixes just use the value of the second function?

fmeum replied:
There were two cycles, the one you linked above and apparently also https://github.com/bazelbuild/bazel/pull/24022#issuecomment-2419309453. The latter goes through the current repo's repo mapping, which may be the main repo mapping.

Wyverald replied:
oh, I missed that other PR... annoying, indeed.

Could we avoid this whole dance by having a version of package lookup that doesn't care about ignored packages? Might not be the best idea... let me think about it more.

lberki replied:
FWIW, I think the current approach is better than  two versions of the package lookup because that one goes much deeper into the internals of Blaze and also much more confusing because it introduces another way to lookup packages that's almost, but not quite, the same as the "regular" one.

The reason why I like the two-phase approach is that it's conceptually sound: one phase parses `REPO.bazel`, then another digests its labels. Yes, it adds some complications, but it's the the cleanest one, at least architecturally.  

Once `WORKSPACE` is gone, we can delete this again (if I understand it correctly). Is that enough consolation?

Wyverald replied:
we actually can't delete this even when WORKSPACE is gone, because computing repo mappings in Bzlmod can also require package lookup first (e.g. `archive_override` with a patch). But I agree that the current approach is the best we can do.

lberki commented:
I addressed the review comments and although the presubmits haven't quite passed yet, they look good so far so I'll context switch away from this. PTAL!

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/PackageArgsFunction.java:
```diff
@@ -0,0 +1,124 @@
+    return new Key(repoName);
+  }
+
+  /** Key type for {@link RepoFileValue}. */
```
PackageArgsValue

lberki replied:
...or RepoPackageArgsValue to be compatible with your other suggestion. Done.

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/PackageArgsFunction.java:
```diff
@@ -0,0 +1,124 @@
+import net.starlark.java.eval.EvalException;
+
+/** A @{link SkyFunction} that returns the {@link PackageArgs} for a given repository. */
+public class PackageArgsFunction implements SkyFunction {
```
This should probably be `RepoPackageArgsFunction`. `PackageArgs` is just the arguments to a `package()` or `repo()` call.

lberki replied:
Done.

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/PackageArgsFunction.java:
```diff
@@ -0,0 +1,124 @@
+
+    @Override
+    public SkyFunctionName functionName() {
+      return SkyFunctions.PACKAGE_ARGS;
```
same here -- rename the SkyFunctionName

lberki replied:
Done

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/PackageArgsFunction.java:
```diff
@@ -0,0 +1,124 @@
+    if (env.valuesMissing()) {
+      return null;
+    }
+    String repoDisplayName = RepoFileFunction.getDisplayNameForRepo(repositoryName, null);
```
why not grab the main repo mapping here and use non-null?

lberki replied:
Done.

Wyverald commented on src/main/java/com/google/devtools/build/lib/packages/RepoThreadContext.java:
```diff
@@ -14,18 +14,26 @@
 
 /** Context object for a Starlark thread evaluating the REPO.bazel file. */
 public class RepoThreadContext extends StarlarkThreadContext {
   private final LabelConverter labelConverter;
```
this isn't used anymore right? we should remove it if so.

Wyverald replied:
double checking: is this done?

lberki replied:
Now it is, sorry. For some reason I thought that you meant unused import statements, saw nothing gray in the IDE and called it good.

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/RepoFileFunction.java:
```diff
@@ -140,7 +136,7 @@ private static String getDisplayNameForRepo(
+  private RepoFileValue evalRepoFile(
       StarlarkFile starlarkFile,
       RepositoryName repoName,
       RepositoryMapping repoMapping,
```
remove the unused repo mapping args

lberki replied:
They are not used (see line 161, the `LabelConverter` call)

Wyverald replied:
assuming you meant "they are not _unused_", they should be now.

Wyverald replied:
ping -- this arg should be removed

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/PackageArgsFunction.java:
```diff
@@ -0,0 +1,124 @@
+      }
+    } catch (EvalException e) {
+      env.getListener().handle(Event.error(e.getMessageWithStack()));
+      throw new RepoFileFunctionException(
```
let's stick with the idiom where we create an exception type for this skyfunction specifically, and keep RepoFileFunctionException ctors private?

lberki replied:
Done. It required more boilerplate than I'd like, but what can I do, this is Java...

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/IgnoredPackagePrefixesFunction.java:
```diff
@@ -44,9 +46,12 @@
  */
 public class IgnoredPackagePrefixesFunction implements SkyFunction {
```
optional: the class name is technically no longer accurate, as this is more "ignored package prefixes and patterns function". Maybe just rename to "IgnoredPackagesFunction", although I understand that'll be a lot more changes across this PR, so up to you.

either way, please update the javadoc a bit to mention it also deals with glob patterns in REPO.bazel files (in addition to .bazelignore).

lberki replied:
If you don't mind, I don't do the rename now because it would indeed be too intrusive. If I don't forget, I'll do it in a followup change, though. I changed the Javadoc, though.

Wyverald commented on src/main/java/com/google/devtools/build/lib/cmdline/IgnoredSubdirectories.java:
```diff
@@ -14,57 +14,123 @@
   public IgnoredSubdirectories filterForDirectory(PathFragment directory) {
     ImmutableSet<PathFragment> filteredPrefixes =
         prefixes.stream().filter(p -> p.startsWith(directory)).collect(toImmutableSet());
+    ImmutableList.Builder<String> filteredPatterns = ImmutableList.builder();
```
were you planning to do something here? if not, just use ImmutableList.of() below

lberki replied:
Argh, that's embarrassing. I thought I wrote a test battery for `IgnoredSubdirectories`, but I either did not or managed to lose it somewhere. Anyways, I wrote one now and fixed this embarrassing mistake.

Wyverald commented on src/main/java/com/google/devtools/build/lib/packages/RepoFileGlobals.java:
```diff
@@ -16,44 +16,62 @@
+  @StarlarkMethod(
+      name = "ignore_directories",
+      useStarlarkThread = true,
+      documented = false, // TODO
```
unTODO please :)

Wyverald replied:
actually, this file doesn't even show up in docs yet (https://bazel.build/rules/lib/globals). The reason was that the `repo()` function is actually special and was documented separately in the build encyclopedia (since its args are "dynamic"). I'll send a follow-up change later to add this to docs. (but please still write the docstring)

lberki replied:
Done!

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/BazelSkyframeExecutorConstants.java:
```diff
@@ -40,7 +42,14 @@ private BazelSkyframeExecutorConstants() {}
    */
   public static final SkyFunction IGNORED_PACKAGE_PREFIXES_FUNCTION =
-      new IgnoredPackagePrefixesFunction(PathFragment.create(".bazelignore"));
+      new IgnoredPackagePrefixesFunction(BAZELIGNORE_PATH, true);
```
IMO it would be better to merge these two args -- or even better, for the NOOP version, just have an inline SkyFunction that returns an empty object. So the non-NOOP version doesn't need any ctor args. `BAZELIGNORE_PATH` could even just be moved to IgnorePackagePrefixesFunction.

lberki replied:
Good idea. Done. This made `IgnoredPackagePrefixesFunction` slightly nicer.

Wyverald reviewed (commented):
LGTM overall

lberki commented:
Comments addressed and tests are green. PTAL!

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/BUILD:
```diff
@@ -2310,6 +2313,25 @@ java_library(
 )
 
+java_library(
+    name = "package_args_function",
```
rename target

lberki replied:
done

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/BazelSkyframeExecutorConstants.java:
```diff
@@ -29,6 +29,8 @@ private BazelSkyframeExecutorConstants() {}
   public static final ImmutableSet<PathFragment> HARDCODED_IGNORED_PACKAGE_PREFIXES =
       ImmutableSet.of();
 
+  public static final PathFragment BAZELIGNORE_PATH = PathFragment.create(".bazelignore");
```
this constant can be inlined -- no reason for it to be a ctor argument, nor a constant in a separate location from its only usage

lberki replied:
Actually, it was used in two places, but it could be fixed, so you ended up being right :) Done.

lberki commented:
Done and done. PTYAL!

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/RepoFileFunction.java:
```diff
@@ -131,7 +126,7 @@ private static StarlarkFile readAndParseRepoFile(Path path, Environment env)
   }
 
-  private static String getDisplayNameForRepo(
+  public static String getDisplayNameForRepo(
```
IMO we should just get rid of this method now that the `mainRepoMapping` arg is always null (in this class). The current caller should just use `repoName.toString()`. Then we can move this method to its only remaining call site.

lberki replied:
If you don't mind, I won't do this in this change; it looks like a subtle behavior change that might make a difference and thus it makes more sense in a separate change.

Wyverald replied:
SGTM!

Wyverald commented on src/main/java/com/google/devtools/build/lib/skyframe/IgnoredPackagePrefixesFunction.java:
```diff
@@ -40,12 +42,23 @@
 
-  public IgnoredPackagePrefixesFunction(PathFragment ignoredPackagePrefixesFile) {
+  public IgnoredPackagePrefixesFunction(
+      PathFragment ignoredPackagePrefixesFile) {
```
sorry to keep harping on about this -- but just in case I didn't get the point through, I don't think this needs to be a ctor argument at all. Its value is only ever `PathFragment.create(".bazelignore")`. Instead of having it passed in by BazelSkyframeExecutorConstants and stored as a member, we can just move the constant into this class and refer to it at lines 132 and 152.

Essentially, I'm approaching this code as if I'm writing it the first time. With that mindset, it's very weird to me that this constant is passed in as a ctor argument. (Yeah I know that it was becuase `NOOP` was implemented using this, but it's no longer done like that.) So I'd like to avoid surprising future readers.

lberki replied:
If the argument was always `.bazelignore`, you were totally right. Sadly, that's not the case at least in three places: `PrepareDepsOfPatternsFunctionSmartNegationTest.java:100`, `SkyframeQueryHelper.createSkyframeExecutor.java:428` and `PackageLookupFunctionTest.java:143`. 

So I'd much rather not rock the boat. Let me see if I have time today to track these down after importing this change...

Wyverald replied:
ah, I see. Those are for testing purposes, so technically still not needed, but I'm happy to leave those for later.

Wyverald reviewed (approved):
few more nits, but overall LGTM! Please import this yourself, as I'd imagine there are a few Blaze-specific changes to be made. Or if you'd rather I do the import, just let me know!
